go/store/datas/database_common.go

```diff
@@ -547,67 +563,186 @@ func (db *database) doUpdateWorkingSet(ctx context.Context, datasetID string, wo
 		return err
 	}
 
-	// This could loop forever, given enough simultaneous writers. BUG 2565
+	workingSetRef, err := db.WriteValue(ctx, workingSet) // will be orphaned if the tryCommitChunks() below fails
+	if err != nil {
+		return err
+	}
+
+	wsValRef, err := types.ToRefOfValue(workingSetRef, db.Format())
+	if err != nil {
+		return err
+	}
+
 	var tryCommitErr error
 	testSetFailed := false
 	for tryCommitErr = ErrOptimisticLockFailed; tryCommitErr == ErrOptimisticLockFailed && !testSetFailed; {
 		tryCommitErr = func() error {
-			db.mu.Lock()
-			defer db.mu.Unlock()
-
 			currentRootHash, err := db.rt.Root(ctx)
+			if err != nil {
+				return err
+			}
 
+			currentDatasets, err := db.DatasetsInRoot(ctx, currentRootHash)
 			if err != nil {
 				return err
 			}
 
-			currentDatasets, err := db.Datasets(ctx)
+			success, err := db.assertDatasetHash(ctx, currentDatasets, datasetID, currHash)
 			if err != nil {
 				return err
 			}
 
-			workingSetRef, err := db.WriteValue(ctx, workingSet) // will be orphaned if the tryCommitChunks() below fails
+			if !success {
+				testSetFailed = true
+				return nil
+			}
+
+			currentDatasets, err = currentDatasets.Edit().Set(types.String(datasetID), wsValRef).Map(ctx)
 			if err != nil {
 				return err
 			}
 
-			ds, err := db.GetDataset(ctx, datasetID)
+			return db.tryCommitChunks(ctx, currentDatasets, currentRootHash)
+		}()
+	}
+
+	return tryCommitErr
+}
+
+// assertDatasetHash returns true if the hash of the dataset matches the one given. Use an empty hash for a dataset you
+// expect not to exist.
+// Typically this is called using optimistic locking by the caller in order to implement atomic test-and-set semantics.
+func (db *database) assertDatasetHash(
+	ctx context.Context,
+	datasets types.Map,
+	datasetID string,
+	currHash hash.Hash,
+) (bool, error) {
+
+	ds, err := db.datasetFromMap(ctx, datasetID, datasets)
+	if err != nil {
+		return false, err
+	}
+
+	if ds.HasHead() {
+		head, ok := ds.MaybeHead()
+		if !ok {
+			panic("no head found")
+		}
+
+		h, err := head.Hash(db.Format())
+		if err != nil {
+			return false, err
+		}
+
+		if h != currHash {
+			return false, err
+		}
+	} else if !currHash.IsEmpty() {
+		return false, nil
+	}
```

Comment:
This seems a tiny bit cleaner (avoids the panic, for example?) as:

```go
if head, ok := ds.MaybeHead(); ok {
    h, err := head.Hash(db.Format())
    if err != nil {
        return false, err
    }
    if h != currHash {
        return false, err
    }
} else if !currHash.IsEmpty() {
    return false, nil
}
```

Reply:
Done